EKF/gps_yaw_fusion.cpp

```diff
@@ -68,91 +69,51 @@ void Ekf::fuseGpsYaw()
-	float t25 = t23+t24;
-	float t26 = 1.0f/t20;
-	float t27 = q1*t2*2.0f;
-	float t28 = t21*t22;
-	float t29 = t28+1.0f;
-	if (fabsf(t29) < 1e-6f) {
-		return;
-	}
-	float t30 = 1.0f/t29;
-	float t31 = q0*t3*2.0f;
-	float t32 = t31-q3*t2*2.0f;
-	float t33 = q3*t3*2.0f;
-	float t34 = q0*t2*2.0f;
-	float t35 = t33+t34;
-
-	float H_YAW[4];
-	H_YAW[0] = (t35/(t11-t2*(t4-q1*q2*2.0f))-t16*t18*t32)/(t18*t21+1.0f);
-	H_YAW[1] = -t30*(t26*(t27-q2*t3*2.0f)+t16*t22*t25);
-	H_YAW[2] = t30*(t25*t26-t16*t22*(t27-q2*t3*2.0f));
-	H_YAW[3] = t30*(t26*t32+t16*t22*t35);
-
-	// using magnetic heading tuning parameter
+	// using magnetic heading process noise
+	// TODO extend interface to use yaw uncertainty provided by GPS if available
 	const float R_YAW = sq(fmaxf(_params.mag_heading_noise, 1.0e-2f));
 
-	// calculate the innovation and define the innovation gate
-	const float innov_gate = math::max(_params.heading_innov_gate, 1.0f);
-	_heading_innov = predicted_hdg - measured_hdg;
-
-	// wrap the innovation to the interval between +-pi
-	_heading_innov = wrap_pi(_heading_innov);
-
-	// Calculate innovation variance and Kalman gains, taking advantage of the fact that only the first 3 elements in H are non zero
-	// calculate the innovation variance
-	float PH[4];
-	_heading_innov_var = R_YAW;
-
-	for (unsigned row = 0; row <= 3; row++) {
-		PH[row] = 0.0f;
-
-		for (uint8_t col = 0; col <= 3; col++) {
-			PH[row] += P(row,col) * H_YAW[col];
-		}
-
-		_heading_innov_var += H_YAW[row] * PH[row];
-	}
-
-	float heading_innov_var_inv;
+	// calculate intermediate variables
+	const float HK0 = sinf(_gps_yaw_offset);
+	const float HK1 = q0*q3;
+	const float HK2 = q1*q2;
+	const float HK3 = 2*HK0*(HK1 - HK2);
+	const float HK4 = cosf(_gps_yaw_offset);
+	const float HK5 = powf(q1, 2);
+	const float HK6 = powf(q2, 2);
+	const float HK7 = powf(q0, 2) - powf(q3, 2);
+	const float HK8 = HK4*(HK5 - HK6 + HK7);
+	const float HK9 = HK3 - HK8;
+	const float HK10 = 1.0F/HK9;
+	const float HK11 = HK4*q0;
+	const float HK12 = HK0*q3;
+	const float HK13 = HK0*(-HK5 + HK6 + HK7) + 2*HK4*(HK1 + HK2);
+	const float HK14 = HK10*HK13;
+	const float HK15 = HK0*q0 + HK4*q3;
+	const float HK16 = HK10*(HK14*(HK11 - HK12) + HK15);
+	const float HK17 = powf(HK13, 2)/powf(HK9, 2) + 1;
+	const float HK18 = 2/HK17;
+	const float HK19 = 1.0F/(-HK3 + HK8);
+	const float HK20 = HK4*q1;
+	const float HK21 = HK0*q2;
+	const float HK22 = HK13*HK19;
+	const float HK23 = HK0*q1 - HK4*q2;
+	const float HK24 = HK19*(HK22*(HK20 + HK21) + HK23);
+	const float HK25 = HK19*(-HK20 - HK21 + HK22*HK23);
+	const float HK26 = HK10*(-HK11 + HK12 + HK14*HK15);
+	const float HK27 = -HK16*P(0,0) - HK24*P(0,1) - HK25*P(0,2) + HK26*P(0,3);
+	const float HK28 = -HK16*P(0,1) - HK24*P(1,1) - HK25*P(1,2) + HK26*P(1,3);
+	const float HK29 = 4/powf(HK17, 2);
+	const float HK30 = -HK16*P(0,2) - HK24*P(1,2) - HK25*P(2,2) + HK26*P(2,3);
+	const float HK31 = -HK16*P(0,3) - HK24*P(1,3) - HK25*P(2,3) + HK26*P(3,3);
+	float HK32;
```

Comment:
No need to declare the variable already here. Can be declared const on initialization.

Reply:
I don't see how that is possible. It needs to be declared in the larger scope and only updated if the inverse is numerically OK.

Reply:
As we directly return inside the if statement, we could declare the variable afterwards.
```
float HK32;
if(bla) {
  HK32 = ...;
  status_ = A;
} else {
   // Do stuff not related to HK32
  return;
}
```
could be 
```
if(!bla) {
   // Do stuff not related to HK32
  return;
}
const float HK32 = ...;
status_ = A;
```